### ISSUE TITLE
[Bug]: FnAPI Java SDK Harness doesn't not update user counters in OnTimer callback functions

### ISSUE THREAD

**Issue by y1chi** [apache/beam#29099]
### What happened?

This can be re-produced with simple DoFn with user counters, the updates in counters in processElement are populated properly to MointoringInfos when sending back ProcessBundleInstructionResponse, however the counters in onTimer callback are not recorded.

```
DoFn<KV<String, Long>, KV<Long, Instant>> fn =
          new DoFn<KV<String, Long>, KV<Long, Instant>>() {

            @TimerId(timerId)
            private final TimerSpec spec = TimerSpecs.timer(TimeDomain.EVENT_TIME);

            final Counter startTimers = Metrics.counter("timers", "started_timers_count");
            final Counter firedTimers = Metrics.counter("timers", "fired_timers_count");

            @ProcessElement
            public void processElement(
                @TimerId(timerId) Timer timer,
                @Timestamp Instant timestamp,
                OutputReceiver<KV<Long, Instant>> r) {
              timer
                  .align(Duration.standardMinutes(1))
                  .offset(Duration.standardSeconds(1))
                  .setRelative();
              LOG.info("started timer");
              startTimers.inc();
              r.output(KV.of(3L, timestamp));
            }

            @OnTimer(timerId)
            public void onTimer(@Timestamp Instant timestamp, OutputReceiver<KV<Long, Instant>> r) {
              LOG.info("fired timer");
              firedTimers.inc();
              r.output(KV.of(42L, timestamp));
            }
          };
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [X] Component: Google Cloud Dataflow Runner

**Comment by chamikaramj** [comment]
Changing to P1 since this sounds like a potential correctness issue.

@y1chi @kennknowles any idea regarding the fix ?

**Comment by kennknowles** [comment]
Agree with P1. I don't immediately know the fix. Was it just forgotten in the implementation?

**Comment by y1chi** [comment]
The Counter logic seems pretty complicated. I suspect there is a missing piece in FnApiDoFnRunner to update the counter containers when the onTimer invocation happens https://github.com/apache/beam/blob/ffb43321a0a4a646c6568336b8fba5787b1b64a0/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L1743

**Comment by yardenbm** [comment]
Not 100% sure this is related, but with my pipeline I have a problem where @OnTimer("loopingTimer") is not triggered when upgraded to 2.52.0.
my pipeline is pretty similar in logic to: https://beam.apache.org/blog/looping-timers/ , while it is working on 2.51.0.

**Comment by kennknowles** [comment]
@yardenbm I don't think it is related. This bug is much older than that, and the execution does not depend on counters. Can you file a new bug for your issue so we can get details and track it?

**Comment by liferoad** [comment]
https://github.com/apache/beam/pull/35494 confirmed this should not happen any more.

**Comment by Robbllle** [comment]
@liferoad 
Sorry to resurrect this, I'm currently implementing a custom counter metric that gets incremented when a timer fires.
I tried updating to v2.68.0 from v2.67.0 to check if this was somehow missed in the 2.67.0 release but no dice :(

My pipeline is running on Google Dataflow with:
* Dataflow Prime
* Runner v2
* Streaming Engine
* Streaming Mode: Exactly once

**Comment by liferoad** [comment]
@Robbllle can you open a cloud support ticket for this? So we can get more details about your job.

**Comment by Robbllle** [comment]
Created support case #63818482

**Comment by scwhittle** [comment]
Sorry you're right, this still doesn't work e2e. The unit test added in the above PR avoids the problem by setting up a MetricsContainer override but in non-test cases, the metrics will function only if an ExecutionState has been activated for the transform. When we are consuming input data from the data stream this is taken care of by the PCollectionConsumerRegistry before handing off to the consumer. Such ExecutionState activation isn't performed before handing off timers to the timer consumer.

I prototyped a solution to expose the existing process bundle execution states from PCollectionConsumerRegistry to ProcessBundleHandler and activating them before forwarding timer data onto FnHarnessRunner.processTimer .  This fixed the counters set during timer firing for a reproduction I put together.

I will work on cleaning up the change to get it submitted.  Potential workarounds in the meantime would be to log or to output (using multiple output PCollections) to a subsequent DoFn which could itself increment the counter. Fusion should keep that processing inline anyway with minimal overhead.

**Comment by liferoad** [comment]
b/453624712
